File: web/storagenode/src/storagenode/api/payout.ts
// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    EstimatedPayout,
    PaymentInfoParameters,
    PayoutApi,
    PayoutPeriod,
    Paystub,
    PreviousMonthEstimatedPayout,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod, SatellitePricingModel,
} from '@/storagenode/payouts/payouts';
import { HttpClient } from '@/storagenode/utils/httpClient';

/**
 * PayoutHttpApi is a http implementation of Payout API.
 * Exposes all payout-related functionality
 */
export class PayoutHttpApi implements PayoutApi {
    private readonly client: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/heldamount';

    /**
     * Fetch paystubs for selected period.
     *
     * @returns paystubs for given period
     * @throws Error
     */
    public async getPaystubsForPeriod(paymentInfoParameters: PaymentInfoParameters): Promise<Paystub[]> {
        let path = `${this.ROOT_PATH}/paystubs/`;

        if (paymentInfoParameters.start) {
            path += paymentInfoParameters.start.period + '/';
        }

        path += paymentInfoParameters.end.period;

        if (paymentInfoParameters.satelliteId) {
            path += '?id=' + paymentInfoParameters.satelliteId;
        }

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get held information');
        }

        const responseBody = await response.json() || [];
        const data: any[] = !Array.isArray(responseBody) ? [ responseBody ] : responseBody; // eslint-disable-line @typescript-eslint/no-explicit-any

        return data.map((paystubJson: any) => { // eslint-disable-line @typescript-eslint/no-explicit-any
            return new Paystub(
                paystubJson.usageAtRest,
                paystubJson.usageGet,
                paystubJson.usagePut,
                paystubJson.usageGetRepair,
                paystubJson.usagePutRepair,
                paystubJson.usageGetAudit,
                paystubJson.compAtRest,
                paystubJson.compGet,
                paystubJson.compPut,
                paystubJson.compGetRepair,
                paystubJson.compPutRepair,
                paystubJson.compGetAudit,
                paystubJson.surgePercent,
                paystubJson.held,
                paystubJson.owed,
                paystubJson.disposed,
                paystubJson.paid,
                paystubJson.distributed,
            );
        });
    }

    /**
     * Fetches available payout periods.
     *
     * @returns payout periods list
     * @throws Error
     */
    public async getPayoutPeriods(id: string): Promise<PayoutPeriod[]> {
        let path = `${this.ROOT_PATH}/periods`;

        if (id) {
            path += '?id=' + id;
        }

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get payout periods');
        }

        const result = await response.json() || [];

        return result.map(period => {
            return PayoutPeriod.fromString(period);
        });
    }

    /**
     * Fetch payout history for given period.
     *
     * @returns payout information
     * @throws Error
     */
    public async getPayoutHistory(period: string): Promise<SatellitePayoutForPeriod[]> {
        const path = `${this.ROOT_PATH}/payout-history/${period}`;

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get payout history information');
        }

        const data: any = await response.json() || []; // eslint-disable-line @typescript-eslint/no-explicit-any

        return data.map((payoutHistoryItem: any) => { // eslint-disable-line @typescript-eslint/no-explicit-any
            return new SatellitePayoutForPeriod(
                payoutHistoryItem.satelliteID,
                payoutHistoryItem.satelliteURL,
                payoutHistoryItem.age,
                payoutHistoryItem.earned,
                payoutHistoryItem.surge,
                payoutHistoryItem.surgePercent,
                payoutHistoryItem.held,
                payoutHistoryItem.afterHeld,
                payoutHistoryItem.disposed,
                payoutHistoryItem.paid,
                payoutHistoryItem.receipt,
                payoutHistoryItem.isExitComplete,
                payoutHistoryItem.heldPercent,
                payoutHistoryItem.distributed,
            );
        });
    }

    /**
     * Fetch total payout information.
     *
     * @returns total payout information
     * @throws Error
     */
    public async getHeldHistory(): Promise<SatelliteHeldHistory[]> {
        const path = `${this.ROOT_PATH}/held-history`;

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get held history information');
        }

        const data: any = await response.json() || []; // eslint-disable-line @typescript-eslint/no-explicit-any

        return data.map((historyItem: any) => { // eslint-disable-line @typescript-eslint/no-explicit-any
            return new SatelliteHeldHistory(
                historyItem.satelliteID,
                historyItem.satelliteName,
                historyItem.holdForFirstPeriod,
                historyItem.holdForSecondPeriod,
                historyItem.holdForThirdPeriod,
                historyItem.totalHeld,
                historyItem.totalDisposed,
                new Date(historyItem.joinedAt),
            );
        });
    }

    /**
     * Fetch estimated payout information.
     *
     * @returns estimated payout information
     * @throws Error
     */
    public async getEstimatedPayout(satelliteId: string): Promise<EstimatedPayout> {
        let path = '/api/sno/estimated-payout';

        if (satelliteId) {
            path += '?id=' + satelliteId;
        }

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get estimated payout information');
        }

        const data: any = await response.json() || new EstimatedPayout(); // eslint-disable-line @typescript-eslint/no-explicit-any

        return new EstimatedPayout(
            new PreviousMonthEstimatedPayout(
                data.currentMonth.egressBandwidth,
                data.currentMonth.egressBandwidthPayout,
                data.currentMonth.egressRepairAudit,
                data.currentMonth.egressRepairAuditPayout,
                data.currentMonth.diskSpace,
                data.currentMonth.diskSpacePayout,
                data.currentMonth.heldRate,
                data.currentMonth.payout,
                data.currentMonth.held,
            ),
            new PreviousMonthEstimatedPayout(
                data.previousMonth.egressBandwidth,
                data.previousMonth.egressBandwidthPayout,
                data.previousMonth.egressRepairAudit,
                data.previousMonth.egressRepairAuditPayout,
                data.previousMonth.diskSpace,
                data.previousMonth.diskSpacePayout,
                data.previousMonth.heldRate,
                data.previousMonth.payout,
                data.previousMonth.held,
            ),
            data.currentMonthExpectations,
        );
    }

    public async getPricingModel(satelliteId: string): Promise<SatellitePricingModel> {
        if (!satelliteId) {
            return new SatellitePricingModel();
        }

        const path = '/api/sno/satellites/'+ satelliteId +'/pricing';

        const response = await this.client.get(path);

        if (!response.ok) {
            throw new Error('can not get satellite pricing information');
        }

        const data: any = await response.json() || new SatellitePricingModel(); // eslint-disable-line @typescript-eslint/no-explicit-any

        return new SatellitePricingModel(
            data.satelliteID,
            data.egressBandwidth,
            data.repairBandwidth,
            data.auditBandwidth,
            data.diskSpace,
        );
    }
}

File: satellite/snopayout/payout.go
// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

package snopayout

import (
	"context"
	"time"

	"github.com/zeebo/errs"
	"go.uber.org/zap"

	"storj.io/common/storj"
)

// DB exposes all needed functionality to manage payout.
//
// architecture: Service
type DB interface {
	// GetPaystub return payStub by nodeID and period.
	GetPaystub(ctx context.Context, nodeID storj.NodeID, period string) (PayStub, error)
	// GetAllPaystubs return all payStubs by nodeID.
	GetAllPaystubs(ctx context.Context, nodeID storj.NodeID) ([]PayStub, error)
	// CreatePaystub insert paystub into db.
	CreatePaystub(ctx context.Context, stub PayStub) (err error)
	// GetPayment return storagenode payment by nodeID and period.
	GetPayment(ctx context.Context, nodeID storj.NodeID, period string) (StoragenodePayment, error)
	// CreatePayment insert payment into db.
	CreatePayment(ctx context.Context, payment StoragenodePayment) (err error)
	// GetAllPayments return all payments by nodeID.
	GetAllPayments(ctx context.Context, nodeID storj.NodeID) ([]StoragenodePayment, error)
}

// ErrNoDataForPeriod represents errors from the payout database.
var ErrNoDataForPeriod = errs.Class("no payStub/payments for period error")

// Error is the default error class for payout package.
var Error = errs.Class("payout db error")

// PayStub is an entity that holds held amount of cash that will be paid to storagenode operator after some period.
type PayStub struct {
	Period         string       `json:"period"`
	NodeID         storj.NodeID `json:"nodeId"`
	Created        time.Time    `json:"created"`
	Codes          string       `json:"codes"`
	UsageAtRest    float64      `json:"usageAtRest"`
	UsageGet       int64        `json:"usageGet"`
	UsagePut       int64        `json:"usagePut"`
	UsageGetRepair int64        `json:"usageGetRepair"`
	UsagePutRepair int64        `json:"usagePutRepair"`
	UsageGetAudit  int64        `json:"usageGetAudit"`
	CompAtRest     int64        `json:"compAtRest"`
	CompGet        int64        `json:"compGet"`
	CompPut        int64        `json:"compPut"`
	CompGetRepair  int64        `json:"compGetRepair"`
	CompPutRepair  int64        `json:"compPutRepair"`
	CompGetAudit   int64        `json:"compGetAudit"`
	SurgePercent   int64        `json:"surgePercent"`
	Held           int64        `json:"held"`
	Owed           int64        `json:"owed"`
	Disposed       int64        `json:"disposed"`
	Paid           int64        `json:"paid"`
}

// StoragenodePayment is an entity that holds payment to storagenode operator parameters.
type StoragenodePayment struct {
	ID      int64        `json:"id"`
	Created time.Time    `json:"created"`
	NodeID  storj.NodeID `json:"nodeId"`
	Period  string       `json:"period"`
	Amount  int64        `json:"amount"`
	Receipt string       `json:"receipt"`
	Notes   string       `json:"notes"`
}

// Service is used to store and handle node paystub information
//
// architecture: Service
type Service struct {
	log *zap.Logger
	db  DB
}

// NewService returns a new Service.
func NewService(log *zap.Logger, db DB) *Service {
	return &Service{
		log: log,
		db:  db,
	}
}

// GetPayStub returns PayStub by nodeID and period.
func (service *Service) GetPayStub(ctx context.Context, nodeID storj.NodeID, period string) (PayStub, error) {
	payStub, err := service.db.GetPaystub(ctx, nodeID, period)
	if err != nil {
		return PayStub{}, Error.Wrap(err)
	}

	return payStub, nil
}

// GetAllPaystubs returns all paystubs by nodeID.
func (service *Service) GetAllPaystubs(ctx context.Context, nodeID storj.NodeID) ([]PayStub, error) {
	payStubs, err := service.db.GetAllPaystubs(ctx, nodeID)
	if err != nil {
		return []PayStub{}, Error.Wrap(err)
	}

	return payStubs, nil
}

// GetPayment returns storagenode payment data by nodeID and period.
func (service *Service) GetPayment(ctx context.Context, nodeID storj.NodeID, period string) (StoragenodePayment, error) {
	payment, err := service.db.GetPayment(ctx, nodeID, period)
	if err != nil {
		return StoragenodePayment{}, Error.Wrap(err)
	}

	return payment, nil
}

// GetAllPayments returns all payments by nodeID.
func (service *Service) GetAllPayments(ctx context.Context, nodeID storj.NodeID) ([]StoragenodePayment, error) {
	payments, err := service.db.GetAllPayments(ctx, nodeID)
	if err != nil {
		return nil, Error.Wrap(err)
	}

	return payments, nil
}
